### ISSUE TITLE
Odoo hardcodes number of proxies in front of it -- need a config option to change it

### ISSUE THREAD

**Issue by lelutin** [odoo/odoo#139536]
Impacted versions: v16, v15, most probably older too

Steps to reproduce: Setup two proxies in front of the wsgi app, configure each level to send X-Forwarded-* headers (e.g. here with nginx):

~~~
proxy_set_header                X-Forwarded-Host $host:$server_port;
proxy_set_header                X-Forwarded-Server $host;
proxy_set_header                X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header                X-Forwarded-Proto $scheme;
proxy_set_header                Host $host;
~~~

Current behavior:

The IP of the last proxy shows up in odoo's log file instead of the client's IP address.

I believe that this might also impact modules that need to use client's IP addresses directly like "link tracker". Currently running tests with regards to this.

Odoo currently hardcodes the number of trusted proxies in front of it when it's set in proxy_mode:

https://github.com/odoo/odoo/blob/71db47cff99ce3b2090c3be5dfa3205654625d51/odoo/http.py#L153C61-L153C61

Expected behavior:

It would be nice to have a positive-integer configuration option, e.g. proxy_num_trusted with a default of 1, that would let odoo admins set how many proxies should be trusted in werkzeug's ProxyFix

With a setup as described above, if I modify the line that creates the ProxyFix class to set all arguments to 2 and restart odoo, I start seeing client IP addresses in odoo's log file.

Video/Screenshot link (optional):

Support ticket number submitted via odoo.com/help (optional):

**Comment by vava-odoo** [comment]
@Julien00859, any idea on this?

**Comment by Julien00859** [comment]
Rejected PR: https://github.com/odoo/odoo/pull/118629
Prior discussion: https://github.com/odoo/odoo/issues/104947#issuecomment-1505552791

TL;DR: You have to configure your web server so that the last entry inside the x-forwarded-for list is the user's ip address. You must not add the various IP address of your various trusted proxies into that list.

**Comment by lelutin** [comment]
@Julien00859 I see, thanks for the feedback on this and links to the other discussions about this.

In that case, it might be good to add some details to odoo's CLI documentation to make it clear that odoo only trusts one level of proxy and thus that any additional layer should not modify the header, and why (is it because information coming from additional layers is hard to trust? the reason for this limitation was not super clear for me in the other issue)

https://www.odoo.com/documentation/16.0/developer/reference/cli.html#cmdoption-odoo-bin-proxy-mode

as a side,note the URL for the `--proxy-mode` option that points to the werkzeug documentation is a dead link. is odoo still using the 0.16 library? the current version of werkzeug is 3.0.x

**Comment by Julien00859** [comment]
@lelutin I agree with you regarding the lack of documentation, this is a recurring question from customers, I'll see what I can do.

The reason why Odoo only take the last value from the `X-Forwarded-For` list is historical, when the `--proxy-mode` was first introduced, the `ProxyFix` middleware class of werkzeug was only reading that last value and it was not configurable. It has since then been changed, the middleware can now be configured to read the last *n-th* value from the list. This is the configuration I used inside my PR that got rejected.

Actually always reading the last value is not that bad, take the following (not so rare) network architecture:

![proxymode drawio](https://github.com/odoo/odoo/assets/8208953/5c506d2d-b607-4df2-af53-2db474023611)

* Customers access Odoo via cloudflare and then via nginx. There are 2 IPs inside the `X-Forwarded-For` list: the one of the customer and then the one of cloudflare. Cloudflare is an active equipment in regard of HTTP.
* Employees access Odoo via a vpn and then via nginx. There is a 1 IP inside the `X-Forwarded-For` list: the one of the employee. A VPN is a passive equipment in regard of HTTP.

We cannot always take the last header as this would use the wrong IP for the requests of customers. We cannot always take the second last header as this would break for the requests of employees.

What we should do instead is to list all the IPs that we should ignore inside the `X-Forwarded-For` list, in this example the IP address of cloudflare and keep on using the last IP that is not ignored. Such a feature already exist in nginx as the (`set_real_ip_from`)[https://nginx.org/en/docs/http/ngx_http_realip_module.html] directive. Because it exists inside nginx, we decided that it was useless to replicate the feature inside Odoo.

About the version of werkzeug, we are compatible with 0.16, 2.0 and 2.2. When using the `.deb` package, it is the `python3-werkzeug` package found in your distribution that is in used ([debian](https://packages.debian.org/search?keywords=python3-werkzeug), [ubuntu](https://packages.ubuntu.com/search?keywords=python3-werkzeug)). When using `pip` and the `requirements.txt`, it depends on the Odoo version and your python version, here for Odoo 16.0: https://github.com/odoo/odoo/blob/1722e62193d0b587e508d30b1d460498c1ea4c86/requirements.txt#L58-L59

We'll make Odoo compatible with werkzeug 3.0 when either debian or ubuntu release a new version that ships werkzeug 3.0.
